### PR TITLE
Emit events when a marker is found and lost, when using aframe's <a-marker>

### PR DESCRIPTION
**Pull request by nikolaymihaylov**
**What kind of change does this PR introduce?**

New feature

**Can it be referenced to an Issue? If so what is the issue # ?**

#217

**How can we test it?**

Check the `aframe/examples/marker-events.html` that I created.

**Summary**

When an `<a-marker>` is found or lost, there is currently no event emitted (as far as I know), to "the outside world". In other words, currently an application that is using AR.js cannot react to the marker becoming visible or invisible.

With my addition, `markerFound` and `markerLost`  events will be emitted when the marker changes its `visible` state. The events are only emitted for `<a-marker>` elements which have `emitevents='true'` set.  `emitevents` defaults to `false`, so there will be no additional events emitted for any of the existing users of the AR.js library.

**Does this PR introduce a breaking change?**

No.

**Other information**

~~I also created a new build, and updated some `README` and `CHANGELOG` files. If these should be excluded from the pull request, let me know and I could make the adjustments.~~
EDIT: The build files are no longer included in this pull request.

### CODE REVIEW

**Comment by Angelo8828** [comment]
What I really need is this one. Referencing #294

**Comment by peterlunglum** [comment]
This is great! How can I download this branch to use in my project?

**Comment by nikolaymihaylov** [comment]
Hey @peterlunglum,

You can download my forked version `git@github.com:nikolaymihaylov/AR.js.git` and then do a `git checkout tags/1.5.2` for the changes that are currently present in this pull request.

Keep in mind that this pull request has already diverged a lot with the original repository, so by choosing to use my forked version, you won't be getting other improvements or bugfixes for AR.js.

The proper solution would be to get this pull request merged to the main repository. I do not know what its maintainers think about my changes though.

**Comment by AndiDomi** [comment]
@nikolaymihaylov A really great job! I just tried it and it works really well with `Hiro` marker and `Kanji`, but I cannot make it work with a custom marker for example like

```
<a-marker preset="custom" type="pattern" url="./332A.patt" id="332-A" registerevents>
</a-marker>
```
Am I doing something wrong or is just that custom markers are not supported on your branch?

**Comment by nikolaymihaylov** [comment]
Hey, there is an ongoing effort to sync this branch with the latest `dev` at: https://github.com/nikolaymihaylov/AR.js/pull/1 Thanks @MaesterChestnut for the effort!

We have an issue with it though. If someone has time to take a look, your help is appreciated!

**Comment by leeprobert** [comment]
+1 on this :-)

**Comment by nicolocarpignoli** [comment]
@nikolaymihaylov can you please resolve conflicts first?

**Comment by MaesterChestnut** [comment]
@nikolaymihaylov the PR I have to your branch (which does not have these merge conflicts) now works as expected. If you merge [my PR](https://github.com/nikolaymihaylov/AR.js/pull/1) to your branch, this PR should be ready for merging

**Comment by nicolocarpignoli** [comment]
@MaesterChestnut and @nikolaymihaylov this is a big feature in my opinion. What needs to be done to make this ready to merge?

**Comment by nicolocarpignoli** [comment]
it solves https://github.com/jeromeetienne/AR.js/issues/217

**Comment by MaesterChestnut** [comment]
@nikolaymihaylov please my PR into yours or else I'll need to open a separate PR for this feature

**Comment by nikolaymihaylov** [comment]
The branch in this pull request is now brought up-to-date. Thanks again @MaesterChestnut for your effort in the pull request https://github.com/nikolaymihaylov/AR.js/pull/1

I have now excluded the build files from the pull request, which will hopefully make this easier to merge. That means that for anyone using **my** latest `dev` branch, you'll need to run the `make minify` as described in https://github.com/jeromeetienne/AR.js#how-to-release- .

If anyone has based their applications on this pull request, you can still use the tag `1.5.2` that I had created, as described in https://github.com/jeromeetienne/AR.js/pull/303#issuecomment-384413920, which has the old build files. That is, of course, not recommended, since the code is outdated by now.

**Comment by derzu** [comment]
I hope this branch be merged. But while it is not a short solution is:

At your HTML:

`<a-marker preset='hiro' id="amarker" ref="amarker">`

Create a method to verify the marker state:
```
verifyMarker() {
    //var amarker = this.$refs.amarker; // vuejs
    var amarker = document.querySelector("#amarker")
    if(amarker.object3D.visible == true) {
        console.log('marker is visible');
    }
    else {
        console.log('marker is lost');
    }
}
```

At some point start your verification:

`setInterval(this.verifyMarker, 1000); // 1000 means 1 second`

**Comment by LaurenceNairne** [comment]
Any plans to test and merge this pull request? It seems pretty robust and it'll do wonders for making AR.js more useful.

**Comment by nicolocarpignoli** [comment]
HI @nikolaymihaylov, I will test it myself. I only ask you if you can give me a builded version of your AR.js so I can directly import it on my example and test it. Something like: `    <script src="https://jeromeetienne.github.io/AR.js/aframe/build/aframe-ar.js"></script>
` will be great :) Thanks a lot

File: aframe/src/component-anchor.js [hunk 1]
```diff
@@ -34,6 +34,12 @@ AFRAME.registerComponent('arjs-anchor', {
 			type: 'number',
 			default: 0.6,
 		},
+
```
Comment (jeromeetienne) [inline aframe/src/component-anchor.js]: (thanks a lot for your contributions @nikolaymihaylov, im making comments here but you dont have to do it. i can do the modification myself after the merge)

this seems harmless to emit the event. it is a rare event so no performance impact. we can remove this options for the sake of simplicity

File: aframe/src/component-anchor.js [hunk 2]
```diff
@@ -155,7 +161,16 @@ AFRAME.registerComponent('arjs-anchor', {
...
+			var wasVisible = _this.el.object3D.visible
 			_this.el.object3D.visible = this._arAnchor.object3d.visible
+
+			if( _this.data.emitevents ){
+				if( _this.el.object3D.visible && !wasVisible ){
+					_this.el.emit('markerFound')
+				}else if( !_this.el.object3D.visible && wasVisible ){
+					_this.el.emit('markerLost')
+				}
+			}
```
Comment (jeromeetienne) [inline aframe/src/component-anchor.js]: Nice standalone trick! 

what about the 'cameraTransformMatrix'  case tho ?

File: aframe/src/component-anchor.js [hunk 3]
```diff
@@ -212,6 +227,7 @@ AFRAME.registerPrimitive('a-marker', AFRAME.utils.extendDeep({}, AFRAME.primitiv
 		'preset': 'arjs-anchor.preset',
 		'minConfidence': 'arjs-anchor.minConfidence',
 		'markerhelpers': 'arjs-anchor.markerhelpers',
+		'emitevents': 'arjs-anchor.emitevents',
```
Comment (jeromeetienne) [inline aframe/src/component-anchor.js]: to be removed, as seen above

**Pull request merged**